src/operations/update.ts

```diff
@@ -41,107 +46,260 @@ export interface UpdateResult {
+    const options = this.options ?? {};
+    const ordered = typeof options.ordered === 'boolean' ? options.ordered : true;
+    const command: Document = {
+      update: this.ns.collection,
+      updates: this.operations,
+      ordered
+    };
+
+    if (typeof options.bypassDocumentValidation === 'boolean') {
+      command.bypassDocumentValidation = options.bypassDocumentValidation;
+    }
+
+    // If a command is to be explained, we need to reformat the command after
+    // the other command properties are specified.
+    // const explain = Explain.fromOptions(options);
+    // if (explain) {
+    //   command = decorateWithExplain(writeCommand, explain);
+    // }
+
+    if (collationNotSupported(server, options)) {
+      callback(new MongoError(`server ${server.name} does not support collation`));
+      return;
+    }
+
+    const unacknowledgedWrite = this.writeConcern && this.writeConcern.w === 0;
+    if (unacknowledgedWrite || maxWireVersion(server) < 5) {
+      if (this.operations.find((o: Document) => o.hint)) {
+        callback(new MongoError(`servers < 3.4 do not support hint on update`));
+        return;
+      }
+    }
+
+    if (this.explain && maxWireVersion(server) < 3) {
+      callback(new MongoError(`server ${server.name} does not support explain on update`));
+      return;
+    }
+
+    super.executeCommand(server, session, command, callback);
   }
 }
 
 /** @internal */
-export class UpdateOneOperation extends CommandOperation<UpdateResult> {
-  options: UpdateOptions;
-  collection: Collection;
-  filter: Document;
-  update: Document;
-
+export class UpdateOneOperation extends UpdateOperation {
   constructor(collection: Collection, filter: Document, update: Document, options: UpdateOptions) {
-    super(collection, options);
+    super(
+      collection.s.namespace,
+      [makeUpdateOperation(filter, update, { ...options, multi: false })],
+      options
+    );
 
     if (!hasAtomicOperators(update)) {
       throw new TypeError('Update document requires atomic operators');
     }
-
-    this.options = options;
-    this.collection = collection;
-    this.filter = filter;
-    this.update = update;
   }
 
-  execute(server: Server, session: ClientSession, callback: Callback<UpdateResult>): void {
-    const coll = this.collection;
-    const filter = this.filter;
-    const update = this.update;
-    const options = { ...this.options, ...this.bsonOptions, session, multi: false };
+  execute(
+    server: Server,
+    session: ClientSession,
+    callback: Callback<UpdateResult | Document>
+  ): void {
+    super.execute(server, session, (err, res) => {
+      if (err || !res) return callback(err);
+      if (typeof this.explain !== 'undefined') return callback(undefined, res);
+
+      // TODO: copied from `updateDocuments`, not sure if these are needed anymore
+      if (res.code) return callback(new MongoError(res));
+      if (res.writeErrors) return callback(new MongoError(res.writeErrors[0]));
```

Comment:
TODO here or should we have a ticket?

Reply:
I removed the `TODO`, this was actually needed. Someone may consider deduplicating the code in the future, but I think its fine for now